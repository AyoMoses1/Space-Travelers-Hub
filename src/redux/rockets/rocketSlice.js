import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const FETCH_ROCKETS = 'spaceTraveller/fetchRockets';
const BOOK_ROCKETS = 'spaceTraveller/bookRocket';
const CANCEL_ROCKETS = 'spaceTraveller/cancelRocket';

export const fetchRockets = createAsyncThunk(FETCH_ROCKETS, async () => fetch('https://api.spacexdata.com/v4/rockets')
  .then((response) => response.json())
  .then((data) => {
    const rockets = data.map((rocket) => ({
      id: rocket.id,
      name: rocket.name,
      type: rocket.type,
      description: rocket.description,
      flickrImages: rocket.flickr_images[1],
    }));

    return rockets;
  })
  .catch((error) => {
    throw new Error(error);
  }));

const rocketSlice = createSlice({
  name: 'spaceTraveller',
  initialState: { rockets: [], loading: false, refresh: true },
  reducers: {
    bookRocket: (state, action) => {
      const rockets = state.rockets.map((rocket) => {
        if (rocket.id !== action.payload) return rocket;
        return { ...rocket, reserved: true };
      });
      return { ...state, rockets };
    },

    cancelRocket: (state, action) => {
      const rockets = state.rockets.map((rocket) => {
        if (rocket.id !== action.payload) return rocket;
        return { ...rocket, reserved: false };
      });
      return { ...state, rockets };
    },
  },

  extraReducers: {
    [fetchRockets.pending]: (state) => ({ ...state, loading: true }),
    [fetchRockets.fulfilled]: (state, action) => ({
      ...state,
      rockets: action.payload,
      loading: false,
    }),
    [fetchRockets.rejected]: (state) => ({ ...state, loading: false }),
  },
});

export const bookRocket = (id) => ({
  type: BOOK_ROCKETS,
  payload: id,
});

export const cancelRocket = (id) => ({
  type: CANCEL_ROCKETS,
  payload: id,
});

export default rocketSlice.reducer;
