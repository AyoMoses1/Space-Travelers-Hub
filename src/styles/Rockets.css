.card-list {
  display: flex;
  flex-direction: column;
  gap: 12px;
  justify-content: space-around;
}

.card-list:last-child {
  padding-bottom: 48px;
}

.card {
  flex-wrap: wrap;
  flex-direction: row;
}

.card-img {
  height: 100%;
  border-style: none;
  padding: 10px;
  border-radius: 16px;
}

.badge {
  margin-right: 12px;
}
