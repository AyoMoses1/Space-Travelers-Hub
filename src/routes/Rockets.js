import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  Button, Card, Col, Badge,
} from 'react-bootstrap';
import { bookRocket, cancelRocket } from '../redux/rockets/rocketSlice';
import '../styles/Rockets.css';

const Rockets = () => {
  const { rockets } = useSelector((state) => state.rockets);
  const dispatch = useDispatch();

  const bookHandler = (id, reserved) => {
    if (reserved) {
      return dispatch(cancelRocket(id));
    }
    return dispatch(bookRocket(id));
  };

  return (
    <div className="card-list">
      {rockets.map(({
        id,
        name,
        type,
        flickrImages,
        description,
        reserved,
      }) => (
        <Card key={id}>
          <Col md={3} xs={12}>
            <Card.Img src={flickrImages} />
          </Col>
          <Col md={9} xs={12}>
            <Card.Body style={{ padding: '16px' }} className="rocket-container">
              <Card.Title>{name}</Card.Title>
              <Card.Text>
                {type}
              </Card.Text>
              <Card.Text>
                {reserved && (
                  <Badge bg="info">Reserved</Badge>
                )}
                {description}
              </Card.Text>
              <Button
                variant={reserved ? 'outline-secondary' : 'primary'}
                onClick={() => bookHandler(id, reserved)}
              >
                { reserved ? 'Cancel Reservation' : 'Reserve Rocket'}
              </Button>
            </Card.Body>
          </Col>
        </Card>
      ))}
    </div>
  );
};

export default Rockets;
