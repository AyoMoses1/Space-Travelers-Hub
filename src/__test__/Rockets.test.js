import React from 'react';
import renderer from 'react-test-renderer';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import Rockets from '../routes/Rockets';
import rockets from '../redux/rockets/rocketSlice';

const createTestStore = () => {
  const store = configureStore({
    reducer: {
      rockets,
    },
  });

  return store;
};
export default createTestStore;

describe('Snapshot components test', () => {
  let store;
  let component;

  beforeEach(() => {
    store = createTestStore();

    component = renderer.create(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    ).toJSON();
  });

  test('Rockets component test', () => {
    expect(component).toMatchSnapshot();
  });
});
