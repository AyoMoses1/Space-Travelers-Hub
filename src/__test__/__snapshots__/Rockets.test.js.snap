// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot components test Rockets component test 1`] = `
<div
  className="card-list"
/>
`;
